tests/test_script_production.py

```diff
@@ -1356,3 +1361,256 @@
+    org_script_dir: ScriptDirectory
+
+    def _move_revision_file(
+        self,
+        rev_index: int,
+        destination_dir: str,
+        version_path="scripts/versions",
+    ):
+        rev_hash = self.rev[rev_index]
+        script = self.rev_to_script[rev_hash]
+        target_file = self._get_moved_path(
+            rev_index, destination_dir, version_path
+        )
+        os.makedirs(os.path.dirname(target_file), exist_ok=True)
+        shutil.move(script.path, target_file)
+
+    def _get_moved_path(
+        self,
+        rev_index: int,
+        destination_dir: str = "",
+        version_path="scripts/versions",
+    ):
+        rev_hash = self.rev[rev_index]
+        script = self.rev_to_script[rev_hash]
+        file_name = os.path.basename(script.path)
+        target_file = os.path.join(
+            _get_staging_directory(), version_path, destination_dir, file_name
+        )
+        target_file = os.path.realpath(target_file)
+        return target_file
+
+    def _checkTopologicalOrder(self, sd: ScriptDirectory):
+        for rev_id in self.rev:
+            new_script = sd.get_revision(rev_id)
+            assertions.is_not_(new_script, None)
+
+            old_revisions = {
+                r.revision: r
+                for r in self._iterate_revisions(self.org_script_dir, rev_id)
+            }
+            new_revisions = {
+                r.revision: r for r in self._iterate_revisions(sd, rev_id)
+            }
+
+            eq_(len(old_revisions), len(new_revisions))
+            for common_rev_id in set(old_revisions.keys()).union(
+                new_revisions.keys()
+            ):
+                old_rev = old_revisions[common_rev_id]
+                new_rev = new_revisions[common_rev_id]
+
+                eq_(old_rev.revision, new_rev.revision)
+                eq_(old_rev.down_revision, new_rev.down_revision)
+                eq_(old_rev.dependencies, new_rev.dependencies)
+
+    @staticmethod
+    def _iterate_revisions(sd: ScriptDirectory, rev_id: str):
+        yield from sd.revision_map.iterate_revisions(
+            rev_id, "base", inclusive=True, assert_relative_length=False
+        )
+
+
+class RecursiveScriptDirectoryTest(TestBase, RecursiveMixin):
+    def setUp(self):
+        self.env = staging_env()
+        self.cfg = _sqlite_testing_config()
+        self.cfg.set_main_option("recursive_version_locations", "true")
+
+        self.rev = list(three_rev_fixture(self.cfg))
+        self.rev.extend(multi_heads_fixture(self.cfg, *self.rev[0:3]))
+
+        self.org_script_dir = ScriptDirectory.from_config(self.cfg)
+        self.rev_to_script = {
+            script.revision: script
+            for script in self.org_script_dir.walk_revisions()
+        }
+
+    def tearDown(self):
+        clear_staging_env()
+
+    def test_flat_structure(self):
+        assert len(self.rev) == 6
+
+    def test_flat_and_dir_structure(self):
+        self._move_revision_file(0, "dir_1")
```

Comment:
**mike bayer** (zzzeek) wrote:

i see this is part of the setup, but it's a little hard to follow that this is part of the setup which also builds on the three_rev_fixture / multi_heads_fixture.   also just generally we have some newer pytest-style conventions that are IMO easier to follow.


for clarity, can we change the fixture organization to be like this, using @testing.fixture (which is a facade over pytest.fixture):


    @testing.fixture
    def base_rev_fixture(self):
        rev = list(three_rev_fixture(self.cfg))
        rev.extend(multi_heads_fixture(self.cfg, *self.rev[0:3]))
        return rev

    @testing.fixture
    def flat_and_dir_fixture(self, base_rev_fixture):
        # ideally these wouldn't be attached to "self" in the test,
        # but I see _move_revision_file and _get_moved_path are relying
        # on these as instance variables for now, OK
        org_script_dir = self.org_script_dir = ScriptDirectory.from_config(self.cfg)
        rev_to_script = self.rev_to_script = {
                script.revision: script
                for script in org_script_dir.walk_revisions()
            }
        self._move_revision_file(0, "dir_1")
        self._move_revision_file(2, "dir_1")
        self._move_revision_file(4, "dir_2")
        self._move_revision_file(5, "dir_3")

        return org_script_dir

    def test_flat_and_dir_structure(self, flat_and_dir_fixture):
        sd = flat_and_dir_fixture

        # ... rest of test

View this in Gerrit at https://gerrit.sqlalchemy.org/c/sqlalchemy/alembic/+/4457

Reply:
I am not sure if I understood correctly this example. In the code, I extracted only `self._move_revision_file(0, "dir_1")` part into fixtures, because `base_rev_fixture, org_script_dir, rev_to_script` are common for tests in a given class.

I am aware that this part of code is somewhat tangled, so I tried to refactor it moving to common code to a new class `RevisionMover`, but in the end it turns out there was no readability improvement, so I stayed with current version (with fixtures).